### PR TITLE
Added Build System Integration

### PR DESCRIPTION
**Pull request by Eshanatnight**
Added `Cmakelists.txt`, `.clang-format`, `.clang-tidy` and `premake5.lua` .

Because of this .gitignore file was modified to not include the build directory.
and a `.gitattributes` file was added

### CODE REVIEW

**Comment by Eshanatnight** [comment]
forgot to push `premake5.lua`

**Comment by Frityet** [comment]
Why premake?

**Comment by Eshanatnight** [comment]
Actually i also wrote a cmakelists.txt but for some reason it got deleted? 

And as for why premake, after no specific reason I started using premake instead of cmake because i wanted to learn Lua and i thought the fastest way would be to use it

File: CMakelists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	source/sema.h
+	include/cpp2util.h
+)
+
+add_executable(${PROJECT_NAME} ${SRC_FILES} ${HEADER_FILES})
+target_include_directories(${PROJECT_NAME} PUBLIC include)
+
+# Tests are yet to be added
+# not exactly sure on how to do it yet
+# as it is author compuer dependent
```
Comment (JohelEGP) [inline CMakelists.txt:22]:
```suggestion
# as it is author computer dependent
```

**Comment by Eshanatnight** [comment]
Just added the Cmakelists file. Found why it was not being tracked by git, .txt files were in the `.gitignore`

**Review by JohelEGP — CHANGES_REQUESTED** [review]
I find this unsatisfactory because the executable isn't exported. I can only use this by `add_subdirectory` and not `find_package`.

**Comment by Eshanatnight** [comment]
Need to find a way to make the tests 

> I find this unsatisfactory because the executable isn't exported. I can only use this by `add_subdirectory` and not `find_package`.

Then would we not need to make the tests in to a "sub-projects" of sorts?

**Comment by JohelEGP** [comment]
My comment is not in regard to tests, but to consumers of the `cppfront` executable.

**Comment by Eshanatnight** [comment]
> My comment is not in regard to tests, but to consumers of the `cppfront` executable.

Oh I understand You mean you find the CMakeLists File. The executable can easily be copied/moved with a post build command. To a standard location. This should also help in making and running the tests easier, in my opinion.

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,5 @@
-
+/.vs
+/.vscode
```
Comment (Colliflower) [inline .gitignore:2]:
```suggestion
```
I feel like adding editor specific folders to the .gitignore is a bad rabbit hole, you end up adding them for every IDE. This belongs in a user level .gitignore instead.

Comment (Eshanatnight) [reply .gitignore:2]: Yeah you are right. I'll change it then.

**Comment by Eshanatnight** [comment]
I would like to ask for some suggestions regarding the tests. Should I classify them as a sub-project to `cppfront`. I tried running it via the `add_custom_command` but that is resulting in either by failing to find the files or giving a bunch of errors.

**Comment by Colliflower** [comment]
I think it might be better to split some of the content from this PR into multiple PRs btw. There are a few things that seem to be unrelated to build systems such as the .md file changes, clang-format and clang-tidy. Those getting their own PRs would clean things up and make this PR more attractive imo.

**Comment by Eshanatnight** [comment]
I apologise, my brain gets all over the place when I work, so it got a bit out of hand it that way. 

What i thought of about the testing, is that it's better if we just keep the Tests as they are atleast for now.

**Comment by Eshanatnight** [comment]
The CMakeLists is complete, but I am keeping The `premake5.lua` file as imo it creates a better visual studio `sln` file than cmake.

**Comment by filipsajdak** [comment]
@Eshanatnight your PR is a collection of unrelated changes (adding build system files, minor edits in Markdown files) - consider moving them to separate PRs. It will make reviewing them easier.

**Comment by Eshanatnight** [comment]
> @Eshanatnight your PR is a collection of unrelated changes (adding build system files, minor edits in Markdown files) - consider moving them to separate PRs. It will make reviewing them easier.

I apologize, I will remove the misc files and edits and remove the premake file

**Comment by Eshanatnight** [comment]
I have made the asked changes. 
Again apologies for the mis steps.

**Comment by Arthapz** [comment]
I have made the support for XMake https://github.com/TapzCrew/cppfront-xmake

File: CMakelists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# issue with the standard library of MSVC
+# https://stackoverflow.com/questions/64889383/how-to-enable-stdclatest-in-cmake
+set(CMAKE_CXX_STANDARD 23)
+
+set(SRC_FILES
+	"source/cppfront.cpp"
+)
+
+add_executable(${PROJECT_NAME} ${SRC_FILES})
+target_include_directories(${PROJECT_NAME} PUBLIC "include")
```
Comment (JohelEGP) [inline CMakelists.txt:20]: Compiling `cppfront` doesn't require this.

Comment (Eshanatnight) [reply CMakelists.txt:20]: Can you please clarify what is not required?

Comment (JohelEGP) [reply CMakelists.txt:20]: Line 20. You can compile `cppfront` without `cpp2util.h`.

**Comment by hsutter** [comment]
Thanks for this idea and energy.

How about making this a separate companion project, and I can add a wiki page of links to known companion projects to help make people aware of them?

My concern about putting it here is that it complicates what I will need to maintain, particularly at this early stage when a lot of things are still fluid and subject to change. And while I realize CMake is usually the most popular and it's a meta-build system that can generate others, if we add CMake I would expect to get questions like "why not Ninja, msbuild, Xcode, Bazel, Gradle, ...?" which I'd have to answer.

Right now I've been considering "no CMake/Bazel/..." to be a feature, in that the project is simple enough that it's easy for anyone to separately build and integrate with whatever build tools and project system they're using. Is that reasonable for now?

**Comment by JohelEGP** [comment]
Yes, please. I'd like to add https://github.com/JohelEGP/jegp.cmake_modules/#jegpcpp2 to the wiki, which allows to compile CMake targets with Cpp2 source files. You can see it working on CE: https://godbolt.org/z/cYf1MrPM4.

File: CMakelists.txt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.2)
+project(cppfront)
+
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY 					${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY 					${CMAKE_BINARY_DIR}/bin)
```
Comment (alexreinking) [inline CMakelists.txt:5]: This is super unfriendly to FetchContent users and totally unnecessary.

File: CMakelists.txt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.2)
```
Comment (alexreinking) [inline CMakelists.txt:1]: CMake 3.2 predates even C++17, so this minimum is bogus.

File: CMakelists.txt (hunk 4 above)
Comment (Eshanatnight) [reply CMakelists.txt:5]: I am using MSVC cmd line. And the cmd line tool has a tendency to make to obfuscate the build dir imo thats why I added that.

File: CMakelists.txt (hunk 5 above)
Comment (Eshanatnight) [reply CMakelists.txt:1]: I am sorry I had no idea that was the case. Thanks For Letting me know.

**Comment by hsutter** [comment]
> Yes, please. I'd like to add https://github.com/JohelEGP/jegp.cmake_modules/#jegpcpp2 to the wiki, which allows to compile CMake targets with Cpp2 source files. You can see it working on CE: https://godbolt.org/z/e1r3qsE8M.

Done, thanks. Added to [wiki home page](https://github.com/hsutter/cppfront/wiki)

**Pull request closed without merging**